File: LeetCodePractice/src/FindLargestValueInEachTreeRow/[515]在每个树行中找最大值.java
package FindLargestValueInEachTreeRow;//您需要在二叉树的每一行中找到最大的值。
//
// 示例： 
//
// 
//输入: 
//
//          1
//         / \
//        3   2
//       / \   \  
//      5   3   9 
//
//输出: [1, 3, 9]
// 
// Related Topics 树 深度优先搜索 广度优先搜索 
// 👍 87 👎 0


//leetcode submit region begin(Prohibit modification and deletion)

import java.util.ArrayList;
import java.util.List;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public List<Integer> largestValues(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();

        helper(root, res, 0);
        return res;
    }

    private void helper(TreeNode root, List<Integer> res, int level) {
        if (root == null) {
            return;
        }

        if (level == res.size()) {
            res.add(root.val);  // add new value
        } else {
            res.set(level, Math.max(res.get(level), root.val));  // set new max value
        }

        // process logic
        helper(root.left, res, level + 1);
        helper(root.right, res, level + 1);
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: LeetCodePractice/src/FindLargestValueInEachTreeRow/[515]在每个树行中找最大值.py
# 您需要在二叉树的每一行中找到最大的值。 
# 
#  示例： 
# 
#  
# 输入: 
# 
#           1
#          / \
#         3   2
#        / \   \  
#       5   3   9 
# 
# 输出: [1, 3, 9]
#  
#  Related Topics 树 深度优先搜索 广度优先搜索 
#  👍 87 👎 0


# leetcode submit region begin(Prohibit modification and deletion)
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
import collections


class Solution(object):
    def largestValues(self, root):
        """
        used BFS method.
        :type root: TreeNode
        :rtype: List[int]
        """

        result = []

        if not root:
            return result

        queue = collections.deque()
        queue.append(root)

        while queue:
            level_len = len(queue)

            maximum = float("-inf")

            for _ in range(level_len):
                node = queue.popleft()
                maximum = max(maximum, node.val)

                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

            result.append(maximum)

        return result

# leetcode submit region end(Prohibit modification and deletion)
